TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,268 @@
+        workItemAttributes.ParentObject = device;
+
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+
+        // It's possible to get called twice.
+        if ((!pDeviceContext) || pDeviceContext->fSetBlackSuccess) {
+            return STATUS_SUCCESS;
+        }
+
+        WDF_WORKITEM_CONFIG_INIT(&workItemConfig, SetBlackWorkItem);
+
+        status = WdfWorkItemCreate(&workItemConfig,
+            &workItemAttributes,
+            &hWorkItem);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrintEx((DPFLTR_IHVDRIVER_ID,
+                DPFLTR_ERROR_LEVEL,
+                "TailLight: Workitem creation failure 0x%x\n",
+                status));
+            return status; // Maybe better luck next time.
+        }
+
+        auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(hWorkItem);
+        workItemContext->symLink = symLink;
+    }
+
+    WdfWorkItemEnqueue(hWorkItem);
+
+    return status;
+}
+
+static NTSTATUS TryToOpenIoTarget(WDFIOTARGET target, 
+    CONST UNICODE_STRING symLink) {
+
+    PAGED_CODE();
+
+    WDF_IO_TARGET_OPEN_PARAMS   openParams = {};
+    WDF_IO_TARGET_OPEN_PARAMS_INIT_OPEN_BY_NAME(
+        &openParams,
+        &symLink,
+        STANDARD_RIGHTS_ALL);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+    }
+
+    return status;
+}
+
+void SetBlackCompletionRoutine(
+    _In_ WDFREQUEST Request,
+    _In_ WDFIOTARGET Target,
+    _In_ PWDF_REQUEST_COMPLETION_PARAMS Params,
+    _In_ WDFCONTEXT Context)
+{
+    UNREFERENCED_PARAMETER(Target);
+    UNREFERENCED_PARAMETER(Params);
+    UNREFERENCED_PARAMETER(Context);
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    status = WdfRequestGetStatus(Request);
+    KdPrint(("TailLight: %s WdfRequestSend status: 0x%x\n", __func__, status));
+
+    // One-shot and top of stack, so delete.
+    WdfObjectDelete(Request);
+}
+
+VOID SetBlackWorkItem(
+    WDFWORKITEM workItem)
+    /*++
+
+    Routine Description:
+
+        Creates a WDF workitem to do the SetBlack() call after the driver
+        stack has initialized.
+
+    Arguments:
+
+        workItem - Handle to a pre-allocated WDF work item.
+    --*/
+{
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    WDFDEVICE device = static_cast<WDFDEVICE>(WdfWorkItemGetParentObject(workItem));
+    auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(workItem);
+
+    status = SetBlackAsync(device, workItemContext->symLink);
+
+    if (NT_SUCCESS(status)) {
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+        InterlockedIncrement((PLONG)(&pDeviceContext->fSetBlackSuccess));
```

Comment:
Here's the rationale. Let's see if it's valid first. The IoRegisterPlugPlayNotification callback runs in one thread. CreateWorkItemForIoTargetOpenDevice runs in another. These two threads could be running on different processors. Microcode could be halfway through an operation on a processor core. In addition, the samples use Interlocked operations whenever two or more threads are doing a accessing a piece of memory.

Reply:
Please then add a comment to document that IoRegisterPlugPlayNotification & CreateWorkItemForIoTargetOpenDevice can be called simultaneously.

Also, I suggest replacing `InterlockedIncrement` with `InterlockedExchange` since it seems like you're really setting a flag here, and not incrementing a counter. Or am I misunderstanding your code?

Reply:
The `InterlockedIncrement` function expect a 32bit target pointer. The `fSetBlackSuccess` is of type `BOOLEAN` which appear to be a typedef for the 8bit  `UCHAR` type.

Won't the `InterlockedExchange` call above therefore lead to out-of-bounds write, which would corrupt the driver and possibly trigger a security vulnerability?